File: exercicios/aula13/ex08_tipos/tipos.py
"""Exercício 08 — validação de tipos como atributos sintetizados.

Entrada: declarações e expressões, em qualquer ordem, cada uma terminada por
`;`:

    item -> T id { , id } ;   |   expr ;        T -> int | float | bool

As expressões são analisadas pelo parser do exercício 06 (a AST não muda).
Depois, uma passagem separada — `check_expr(node, env)` — calcula o atributo
sintetizado `type` de cada nó a partir do `type` dos filhos. Os tipos ficam
numa tabela à parte (nó -> tipo); a árvore sintática não é alterada.

Regras (política de promoção: int -> float implícito, nunca o contrário):

    literal inteiro / real / true|false     int / float / bool
    identificador                           o tipo declarado (não declarado: erro)
    + - * /     numérico, numérico          float se algum for float; senão int
    %           int, int                    int
    < <= > >=   numérico, numérico          bool
    == !=       numérico, numérico | bool, bool    bool
    && ||       bool, bool                  bool
    - (unário)  numérico                    o mesmo tipo
    ! (unário)  bool                        bool

Operando incompatível gera um diagnóstico com a posição do operador e o nó
recebe o tipo `erro`. Um nó cujo filho já é `erro` também vira `erro`, mas sem
novo diagnóstico — o erro original não se multiplica pela árvore acima.

Saída: cada expressão com o tipo de cada nó, `(+:float 3.0:float 4:int)`.
Códigos de saída distintos: 3 para erro de sintaxe (nada é verificado) e 4
para erro semântico (a AST é impressa mesmo assim).

Uso:  python tipos.py <arquivo | ->
"""

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1] / "ex06_expressoes_minic"))
import expr as E  # noqa: E402
from expr import F  # noqa: E402

TIPOS_DECL = {"KW_INT": "int", "KW_FLOAT": "float", "KW_BOOL": "bool"}
NUMERICOS = ("int", "float")
ERRO = "erro"


class Verificador:
    def __init__(self):
        self.env: dict[str, str] = {}
        self.tipos: dict[int, str] = {}  # id(nó) -> tipo: o atributo, fora da AST
        self.erros: list[str] = []

    def _erro(self, no, mensagem: str):
        self.erros.append(F.erro_semantico(no.linha, no.coluna, mensagem))

    def check_expr(self, no: E.Expr) -> str:
        tipo = self._tipo(no)
        self.tipos[id(no)] = tipo
        return tipo

    def _tipo(self, no: E.Expr) -> str:
        match no:
            case E.Integer():
                return "int"
            case E.Float():
                return "float"
            case E.Bool():
                return "bool"
            case E.Identifier(nome=nome):
                if nome not in self.env:
                    self._erro(no, f"identificador '{nome}' não declarado")
                    return ERRO
                return self.env[nome]
            case E.Unary(op=op, operando=o):
                t = self.check_expr(o)
                if t == ERRO:
                    return ERRO
                if op == "-" and t in NUMERICOS:
                    return t
                if op == "!" and t == "bool":
                    return "bool"
                exigido = "numérico" if op == "-" else "bool"
                self._erro(no, f"operador '{op}' exige operando {exigido}; recebeu {t}")
                return ERRO
            case E.Binary(op=op, esq=e, dir=d):
                a, b = self.check_expr(e), self.check_expr(d)
                if ERRO in (a, b):
                    return ERRO
                return self._binario(no, op, a, b)

    def _binario(self, no, op: str, a: str, b: str) -> str:
        numericos = a in NUMERICOS and b in NUMERICOS
        if op in ("+", "-", "*", "/"):
            if numericos:
                return "float" if "float" in (a, b) else "int"
            exigido = "numéricos"
        elif op == "%":
            if a == b == "int":
                return "int"
            exigido = "int"
        elif op in ("<", "<=", ">", ">="):
            if numericos:
                return "bool"
            exigido = "numéricos"
        elif op in ("==", "!="):
            if numericos or a == b == "bool":
                return "bool"
            exigido = "do mesmo tipo (numéricos ou bool)"
        else:  # && ||
            if a == b == "bool":
                return "bool"
            exigido = "bool"
        self._erro(no, f"operador '{op}' exige operandos {exigido}; recebeu {a} e {b}")
        return ERRO

    def declara(self, tok, tipo: str):
        if tok.lexeme in self.env:
            self.erros.append(
                F.erro_semantico(tok.line, tok.column, f"redeclaração de '{tok.lexeme}'")
            )
            return
        self.env[tok.lexeme] = tipo

    def anotada(self, no: E.Expr) -> str:
        t = self.tipos[id(no)]
        match no:
            case E.Unary(op=op, operando=o):
                return f"({op}:{t} {self.anotada(o)})"
            case E.Binary(op=op, esq=e, dir=d):
                return f"({op}:{t} {self.anotada(e)} {self.anotada(d)})"
            case _:
                return f"{E.sexpr(no)}:{t}"


def main(argv: list[str]) -> int:
    F.configurar_saida()
    if len(argv) != 2:
        print("uso: python tipos.py <arquivo | ->", file=sys.stderr)
        return F.ERRO_USO
    fl, codigo = F.abrir(argv[1])
    if fl is None:
        return codigo

    # 1) análise sintática: declarações e ASTs das expressões
    itens = []
    parser = E.ExprParser(fl)
    try:
        while not fl.verifica("EOF"):
            tok = fl.atual()
            if tok.type in TIPOS_DECL:
                fl.proximo()
                ids = [fl.exige("IDENT", "identificador")]
                while fl.aceita("COMMA"):
                    ids.append(fl.exige("IDENT", "identificador"))
                fl.exige("SEMICOLON", "',' ou ';'")
                itens.append(("decl", TIPOS_DECL[tok.type], ids))
            else:
                arvore = parser.parse_expr()
                fl.exige("SEMICOLON", "operador ou ';'")
                itens.append(("expr", arvore))
    except F.ErroSintaxe as erro:
        print(erro, file=sys.stderr)
        return F.ERRO_SINTAXE

    # 2) análise semântica: atributo `type` de cada nó
    v = Verificador()
    for item in itens:
        if item[0] == "decl":
            for tok in item[2]:
                v.declara(tok, item[1])
        else:
            v.check_expr(item[1])
            print(v.anotada(item[1]))
    for e in v.erros:
        print(e, file=sys.stderr)
    return F.ERRO_SEMANTICO if v.erros else F.OK


if __name__ == "__main__":
    sys.exit(main(sys.argv))

File: exercicios/aula13/ex08_tipos/tipos.c
/* Exercício 08 — validação de tipos como atributos sintetizados.
 *
 * Mesmas regras, mesma política de promoção e mesma saída de tipos.py (ver a
 * docstring de lá). As expressões vêm de ex06_expressoes_minic/expr.c, sem
 * alteração na struct Expr:
 *
 * - check_expr é recursiva e devolve `enum Tipo`;
 * - o atributo de cada nó fica numa tabela à parte (pares nó -> tipo, busca
 *   linear), para não mexer na árvore sintática;
 * - os diagnósticos vão para um vetor de `Diagnostic` e só são impressos no
 *   fim: a verificação nunca encerra o processo.
 *
 * Uso:  ./tipos <arquivo | ->
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "../ex06_expressoes_minic/expr.h"

typedef enum { T_INT, T_FLOAT, T_BOOL, T_ERRO } Tipo;
static const char *NOME_TIPO[] = {"int", "float", "bool", "erro"};

typedef struct {
    int linha, coluna;
    char *mensagem;
} Diagnostic;

typedef struct {
    const Expr *no;
    Tipo tipo;
} Anotacao;

typedef struct {
    char *nome;
    Tipo tipo;
} Var;

typedef struct {
    Var *env;
    size_t n_env, cap_env;
    Anotacao *tipos;
    size_t n_tipos, cap_tipos;
    Diagnostic *diags;
    size_t n_diags, cap_diags;
} Verificador;

static void diag(Verificador *v, int linha, int coluna, char *mensagem) {
    if (v->n_diags == v->cap_diags) {
        v->cap_diags = v->cap_diags ? v->cap_diags * 2 : 8;
        v->diags = realoca(v->diags, v->cap_diags * sizeof(Diagnostic));
    }
    v->diags[v->n_diags++] = (Diagnostic){linha, coluna, mensagem};
}

static const Var *env_procura(const Verificador *v, const char *nome) {
    for (size_t i = 0; i < v->n_env; i++) {
        if (strcmp(v->env[i].nome, nome) == 0) return &v->env[i];
    }
    return NULL;
}

static void declara(Verificador *v, const Token *t, Tipo tipo) {
    if (env_procura(v, t->lexeme)) {
        diag(v, t->line, t->column, formata("redeclaração de '%s'", t->lexeme));
        return;
    }
    if (v->n_env == v->cap_env) {
        v->cap_env = v->cap_env ? v->cap_env * 2 : 8;
        v->env = realoca(v->env, v->cap_env * sizeof(Var));
    }
    v->env[v->n_env++] = (Var){copia(t->lexeme), tipo};
}

static Tipo anota(Verificador *v, const Expr *no, Tipo tipo) {
    if (v->n_tipos == v->cap_tipos) {
        v->cap_tipos = v->cap_tipos ? v->cap_tipos * 2 : 16;
        v->tipos = realoca(v->tipos, v->cap_tipos * sizeof(Anotacao));
    }
    v->tipos[v->n_tipos++] = (Anotacao){no, tipo};
    return tipo;
}

static Tipo tipo_de(const Verificador *v, const Expr *no) {
    for (size_t i = v->n_tipos; i-- > 0;) {
        if (v->tipos[i].no == no) return v->tipos[i].tipo;
    }
    return T_ERRO; /* inalcançável: todo nó é anotado antes de ser impresso */
}

static int numerico(Tipo t) { return t == T_INT || t == T_FLOAT; }

static Tipo binario(Verificador *v, const Expr *no, Tipo a, Tipo b) {
    const char *op = no->texto;
    const char *exigido;
    int nums = numerico(a) && numerico(b);
    if (strchr("+-*/", op[0]) && op[1] == '\0') {
        if (nums) return (a == T_FLOAT || b == T_FLOAT) ? T_FLOAT : T_INT;
        exigido = "numéricos";
    } else if (strcmp(op, "%") == 0) {
        if (a == T_INT && b == T_INT) return T_INT;
        exigido = "int";
    } else if (strcmp(op, "<") == 0 || strcmp(op, "<=") == 0 || strcmp(op, ">") == 0 || strcmp(op, ">=") == 0) {
        if (nums) return T_BOOL;
        exigido = "numéricos";
    } else if (strcmp(op, "==") == 0 || strcmp(op, "!=") == 0) {
        if (nums || (a == T_BOOL && b == T_BOOL)) return T_BOOL;
        exigido = "do mesmo tipo (numéricos ou bool)";
    } else { /* && || */
        if (a == T_BOOL && b == T_BOOL) return T_BOOL;
        exigido = "bool";
    }
    diag(v, no->linha, no->coluna,
         formata("operador '%s' exige operandos %s; recebeu %s e %s", op, exigido, NOME_TIPO[a], NOME_TIPO[b]));
    return T_ERRO;
}

static Tipo check_expr(Verificador *v, const Expr *no) {
    Tipo t = T_ERRO;
    switch (no->k) {
    case E_INT: t = T_INT; break;
    case E_FLOAT: t = T_FLOAT; break;
    case E_BOOL: t = T_BOOL; break;
    case E_IDENT: {
        const Var *var = env_procura(v, no->texto);
        if (var) t = var->tipo;
        else diag(v, no->linha, no->coluna, formata("identificador '%s' não declarado", no->texto));
        break;
    }
    case E_UNARY: {
        Tipo o = check_expr(v, no->a);
        if (o == T_ERRO) break;
        if (no->texto[0] == '-' && numerico(o)) t = o;
        else if (no->texto[0] == '!' && o == T_BOOL) t = T_BOOL;
        else
            diag(v, no->linha, no->coluna,
                 formata("operador '%s' exige operando %s; recebeu %s", no->texto,
                         no->texto[0] == '-' ? "numérico" : "bool", NOME_TIPO[o]));
        break;
    }
    case E_BINARY: {
        Tipo a = check_expr(v, no->a);
        Tipo b = check_expr(v, no->b);
        if (a != T_ERRO && b != T_ERRO) t = binario(v, no, a, b);
        break;
    }
    }
    return anota(v, no, t);
}

static void anotada(const Verificador *v, const Expr *no) {
    const char *t = NOME_TIPO[tipo_de(v, no)];
    switch (no->k) {
    case E_UNARY:
        printf("(%s:%s ", no->texto, t);
        anotada(v, no->a);
        putchar(')');
        break;
    case E_BINARY:
        printf("(%s:%s ", no->texto, t);
        anotada(v, no->a);
        putchar(' ');
        anotada(v, no->b);
        putchar(')');
        break;
    default: printf("%s:%s", no->texto, t);
    }
}

/* Itens lidos na fase sintática: uma declaração (tipo + tokens) ou uma expressão. */
typedef struct {
    int e_decl;
    Tipo tipo;
    const Token **ids;
    size_t n_ids;
    Expr *expr;
} Item;

static Tipo tipo_decl(TokenType t) {
    return t == TOK_KW_INT ? T_INT : t == TOK_KW_FLOAT ? T_FLOAT : t == TOK_KW_BOOL ? T_BOOL : T_ERRO;
}

static const Token *exige(Fluxo *fl, TokenType tipo, const char *esperado) {
    if (fluxo_verifica(fl, tipo)) return fluxo_proximo(fl);
    erro_sintaxe(fluxo_atual(fl), esperado);
    return NULL;
}

/* Fase sintática: preenche *itens; devolve OK ou ERRO_SINTAXE. */
static int analisa(Fluxo *fl, Item **itens, size_t *n) {
    size_t cap = 0;
    while (!fluxo_verifica(fl, TOK_EOF)) {
        if (*n == cap) {
            cap = cap ? cap * 2 : 8;
            *itens = realoca(*itens, cap * sizeof(Item));
        }
        Item *it = &(*itens)[*n];
        *it = (Item){0};
        Tipo td = tipo_decl(fluxo_atual(fl)->type);
        if (td != T_ERRO) {
            fluxo_proximo(fl);
            it->e_decl = 1;
            it->tipo = td;
            (*n)++; /* já conta, para ser liberado mesmo se der erro */
            do {
                const Token *id = exige(fl, TOK_IDENT, "identificador");
                if (!id) return ERRO_SINTAXE;
                it->ids = realoca(it->ids, (it->n_ids + 1) * sizeof(Token *));
                it->ids[it->n_ids++] = id;
            } while (fluxo_aceita(fl, TOK_COMMA));
            if (!exige(fl, TOK_SEMICOLON, "',' ou ';'")) return ERRO_SINTAXE;
        } else {
            ExprErro erro;
            it->expr = expr_parse(fl, &erro);
            if (!it->expr) {
                erro_sintaxe(erro.token, erro.esperado);
                return ERRO_SINTAXE;
            }
            (*n)++;
            if (!exige(fl, TOK_SEMICOLON, "operador ou ';'")) return ERRO_SINTAXE;
        }
    }
    return OK;
}

int main(int argc, char **argv) {
    if (argc != 2) {
        fprintf(stderr, "uso: %s <arquivo | ->\n", argv[0]);
        return ERRO_USO;
    }
    Fluxo fl;
    int codigo = fluxo_abrir(&fl, argv[1]);
    if (codigo != OK) return codigo;

    Item *itens = NULL;
    size_t n = 0;
    codigo = analisa(&fl, &itens, &n);

    Verificador v = {0};
    if (codigo == OK) {
        for (size_t i = 0; i < n; i++) {
            if (itens[i].e_decl) {
                for (size_t k = 0; k < itens[i].n_ids; k++) declara(&v, itens[i].ids[k], itens[i].tipo);
            } else {
                check_expr(&v, itens[i].expr);
                anotada(&v, itens[i].expr);
                putchar('\n');
            }
        }
        for (size_t i = 0; i < v.n_diags; i++) {
            fprintf(stderr, "Erro semântico na linha %d, coluna %d: %s.\n", v.diags[i].linha, v.diags[i].coluna,
                    v.diags[i].mensagem);
        }
        codigo = v.n_diags ? ERRO_SEMANTICO : OK;
    }

    for (size_t i = 0; i < n; i++) free(itens[i].ids), expr_free(itens[i].expr);
    free(itens);
    for (size_t i = 0; i < v.n_env; i++) free(v.env[i].nome);
    for (size_t i = 0; i < v.n_diags; i++) free(v.diags[i].mensagem);
    free(v.env), free(v.tipos), free(v.diags);
    fluxo_fechar(&fl);
    return codigo;
}
